### ISSUE TITLE
All training scripts might be wrong when using gradients accumulation!

### ISSUE THREAD

**Issue by chenbinghui1** [huggingface/diffusers#10553]
Here is a simple case:
```
loss = loss.mean()
accelerator.backward(loss)

if accelerator.sync_gradients:
        params_to_clip = flux_transformer.parameters()
        accelerator.clip_grad_norm_(params_to_clip, args.max_grad_norm)
optimizer.step()
lr_scheduler.step()
optimizer.zero_grad()
```

should be updated to 
```
loss = loss.mean()
accelerator.backward(loss)

if accelerator.sync_gradients:
    params_to_clip = flux_transformer.parameters()
    accelerator.clip_grad_norm_(params_to_clip, args.max_grad_norm)

    optimizer.step()
    lr_scheduler.step()
    optimizer.zero_grad()
```

@yiyixuxu

**Comment by nitinmukesh** [comment]
Not sure why but both code are same? What changes?

**Comment by bghira** [comment]
this is true @linoytsaban the HF accelerate docs on the subject are [here](https://huggingface.co/docs/accelerate/main/en/basic_tutorials/migration#gradient-accumulation)

**Comment by bghira** [comment]
@nitinmukesh the optimizer step and gradient zero should only occur after they are synced, i think the accumulation ends up merely skipping steps in the current code.

**Comment by yiyixuxu** [comment]
ohh makes sense to me
can you take a look here, @sayakpaul @linoytsaban

**Comment by sayakpaul** [comment]
Thanks for bringing this to our attention, @chenbinghui1! Do you want to submit a PR for the Flux training scripts, maybe, as you have already proposed a fix?

**Comment by sayakpaul** [comment]
Wait, https://huggingface.co/docs/accelerate/en/usage_guides/gradient_accumulation doesn't hint at using the step() calls under `sync_gradients`. Ccing @muellerzr here too.
